iwyu_ast_util.cc

```diff
@@ -538,41 +573,19 @@ string GetWrittenQualifiedNameAsString(const NamedDecl* named_decl) {
   printing_policy.SuppressUnwrittenScope = true;
   named_decl->printQualifiedName(ostream, printing_policy);
 
-  if (const auto* spec =
+  if (const auto* cls_spec =
           dyn_cast<ClassTemplateSpecializationDecl>(named_decl)) {
-    if (spec->isExplicitSpecialization()) {
-      ostream << '<';
-      const TemplateArgumentList& args = spec->getTemplateArgs();
-      const TemplateParameterList* params =
-          spec->getSpecializedTemplate()->getTemplateParameters();
-      for (unsigned i = 0, ie = args.size(); i < ie; ++i) {
-        if (i > 0)
-          ostream << ", ";
-        bool include_type = TemplateParameterList::shouldIncludeTypeForArgument(
-            printing_policy, params, i);
-        const TemplateArgument& arg = args.get(i);
-        // Clang embraces arguments corresponding to a parameter pack in angle
-        // brackets, so cannot use the default 'print' for them.
-        if (arg.getKind() == TemplateArgument::Pack) {
-          ArrayRef<TemplateArgument> pack_args = arg.getPackAsArray();
-          for (size_t j = 0, je = pack_args.size(); j < je; ++j) {
-            if (j > 0)
-              ostream << ", ";
-            pack_args[j].print(printing_policy, ostream, include_type);
-          }
-        } else {
-          args.get(i).print(printing_policy, ostream, include_type);
-        }
-      }
-      ostream << '>';
-
-      // Replace clang placeholders in partial specialization arguments with :N,
-      // where N is the parameter index (depth is ignored for now).
-      static regex tpl_param{"(?:type|value|template)-parameter-\\d+-(\\d+)"};
-      retval = regex_replace(retval, tpl_param, ":$1");
-    }
+    PrintExplicitSpecTplArgs(cls_spec, printing_policy, ostream);
+  } else if (const auto* var_spec =
+                 dyn_cast<VarTemplateSpecializationDecl>(named_decl)) {
+    PrintExplicitSpecTplArgs(var_spec, printing_policy, ostream);
   }
 
+  // Replace clang placeholders in partial specialization arguments with :N,
```

Comment:
Hmm, don't we need to `ostream.flush();` before using `retval` below? Maybe there's never any buffering in `raw_ostring_stream`?

Reply:
Yeah, `there is no need to call flush().`

https://github.com/llvm/llvm-project/blob/4e44e8761795d4672121368929b9c99f6dbaffa7/llvm/include/llvm/Support/raw_ostream.h#L659